File: src/swarm/agents/video.agent.ts
import { BaseAgent, type AgentConfig, registerAgentFactory } from "../base-agent.js";
import { defineTool } from "../llm/tool-executor.js";
import type { AgentContext, ToolDefinition, AgentType } from "../types.js";
import { updateHeyGenVideo } from "../../contacts/index.js";

/**
 * Video Agent Configuration
 */
const AGENT_CONFIG: AgentConfig = {
  type: "video",
  name: "Video Agent",
  description: "Manages HeyGen video generation and coordination.",
  temperature: 0.3, // Low temperature for consistent operations
};

/**
 * VideoAgent - Handles HeyGen video generation
 *
 * Responsibilities:
 * - Initiate video generation requests
 * - Check video generation status
 * - Update contact records with video URLs
 * - Coordinate with personalization agent for scripts
 *
 * Tools:
 * - generate_video: Start HeyGen video generation
 * - check_video_status: Check generation status
 * - get_video_url: Get completed video URL
 */
export class VideoAgent extends BaseAgent {
  constructor() {
    super(AGENT_CONFIG);
  }

  /**
   * Get the system prompt for video operations
   */
  protected getSystemPrompt(context: AgentContext): string {
    const heygenConfigured = !!context.config.heygen?.avatarId;

    return `You are a video generation agent for the networking follow-up system.

## Your Role
Coordinate HeyGen video generation for personalized outreach.

## HeyGen Status
${heygenConfigured ? "HeyGen is configured and ready." : "Warning: HeyGen is not fully configured (missing avatar/voice ID)."}

## Video Generation Process
1. Receive a script (from personalization agent or template)
2. Submit to HeyGen for video generation
3. Monitor generation status
4. Update contact record when complete
5. Notify that video is ready for sending

## Best Practices
- Videos should be 30-60 seconds max
- Scripts should be conversational, not robotic
- Always check if contact already has a video before generating
- Handle generation failures gracefully

## Current Contact
Phone: ${context.phoneNumber || "Unknown"}
${context.contact?.heygen_video_id ? `Existing video ID: ${context.contact.heygen_video_id}` : "No existing video"}
${context.contact?.heygen_video_url ? `Video URL: ${context.contact.heygen_video_url}` : ""}`;
  }

  /**
   * Define tools available to this agent
   */
  protected getTools(): ToolDefinition[] {
    return [
      defineTool(
        "generate_heygen_video",
        "Start generating a HeyGen video with the given script",
        {
          phoneNumber: {
            type: "string",
            description: "Contact's phone number",
          },
          script: {
            type: "string",
            description: "Video script to use",
          },
          webhookUrl: {
            type: "string",
            description: "Optional webhook URL for completion notification",
          },
        },
        ["phoneNumber", "script"]
      ),
      defineTool(
        "check_video_status",
        "Check the status of a HeyGen video generation",
        {
          videoId: {
            type: "string",
            description: "HeyGen video ID",
          },
        },
        ["videoId"]
      ),
      defineTool(
        "get_contact_video",
        "Get the video information for a contact",
        {
          phoneNumber: {
            type: "string",
            description: "Contact's phone number",
          },
        },
        ["phoneNumber"]
      ),
    ];
  }

  /**
   * Register tool handlers
   */
  protected registerTools(): void {
    // Generate HeyGen video
    this.toolExecutor.registerTool(
      defineTool(
        "generate_heygen_video",
        "Generate video",
        {
          phoneNumber: { type: "string" },
          script: { type: "string" },
          webhookUrl: { type: "string" },
        },
        ["phoneNumber", "script"]
      ),
      async (input, context) => {
        const heygen = context.config.heygen;
        if (!heygen?.avatarId || !heygen?.voiceId) {
          return { success: false, error: "HeyGen not configured (missing avatar/voice ID)" };
        }

        const apiKey = process.env.HEYGEN_API_KEY;
        if (!apiKey) {
          return { success: false, error: "HEYGEN_API_KEY not set" };
        }

        try {
          // Build request body with optional webhook callback
          const requestBody: Record<string, unknown> = {
            video_inputs: [
              {
                character: {
                  type: "avatar",
                  avatar_id: heygen.avatarId,
                  avatar_style: "normal",
                },
                voice: {
                  type: "text",
                  voice_id: heygen.voiceId,
                  input_text: input.script as string,
                },
              },
            ],
            dimension: {
              width: 1280,
              height: 720,
            },
            callback_id: input.phoneNumber as string,
          };

          // Add webhook URL if configured - this is essential for automatic video delivery
          // Use explicit webhookUrl from config, or construct from WEBHOOK_BASE_URL env var
          const webhookUrl = heygen.webhookUrl ||
            (process.env.WEBHOOK_BASE_URL ? `${process.env.WEBHOOK_BASE_URL}/webhooks/networking-event/heygen` : null);

          if (webhookUrl) {
            requestBody.callback_url = webhookUrl;
            context.logger.debug({ webhookUrl }, "Webhook URL configured for video callback");
          } else {
            context.logger.warn("No webhook URL configured (set heygen.webhookUrl or WEBHOOK_BASE_URL) - video will not be sent automatically");
          }

          const response = await fetch("https://api.heygen.com/v2/video/generate", {
            method: "POST",
            headers: {
              "Content-Type": "application/json",
              "X-Api-Key": apiKey,
            },
            body: JSON.stringify(requestBody),
          });

          const data = await response.json() as { data?: { video_id: string }; error?: string };

          if (!response.ok || data.error) {
            return {
              success: false,
              error: data.error || `HeyGen API error: ${response.status}`,
            };
          }

          const videoId = data.data?.video_id;

          // Save video_id to contact record so webhook can find them when video is ready
          if (videoId) {
            await updateHeyGenVideo(
              input.phoneNumber as string,
              videoId,
              null, // URL will be set by webhook when video is complete
              context.config.supabase
            );
            context.logger.info(
              { phoneNumber: input.phoneNumber, videoId },
              "Video ID saved to contact - video will be sent automatically when ready"
            );
          }

          return {
            success: true,
            videoId,
            status: "processing",
            message: "Video generation started - will be sent automatically when ready",
          };
        } catch (error) {
          return {
            success: false,
            error: error instanceof Error ? error.message : "Unknown error",
          };
        }
      }
    );

    // Check video status
    this.toolExecutor.registerTool(
      defineTool(
        "check_video_status",
        "Check status",
        { videoId: { type: "string" } },
        ["videoId"]
      ),
      async (input, context) => {
        const apiKey = process.env.HEYGEN_API_KEY;
        if (!apiKey) {
          return { success: false, error: "HEYGEN_API_KEY not set" };
        }

        try {
          const response = await fetch(
            `https://api.heygen.com/v1/video_status.get?video_id=${input.videoId}`,
            {
              headers: { "X-Api-Key": apiKey },
            }
          );

          const data = await response.json() as {
            data?: { status: string; video_url?: string };
            error?: string;
          };

          if (!response.ok || data.error) {
            return {
              success: false,
              error: data.error || `HeyGen API error: ${response.status}`,
            };
          }

          const status = data.data?.status;
          const videoUrl = data.data?.video_url;

          // If completed, update contact record
          if (status === "completed" && videoUrl && context.phoneNumber) {
            await updateHeyGenVideo(
              context.phoneNumber,
              input.videoId as string,
              videoUrl,
              context.config.supabase
            );
          }

          return {
            success: true,
            status,
            videoUrl,
            completed: status === "completed",
          };
        } catch (error) {
          return {
            success: false,
            error: error instanceof Error ? error.message : "Unknown error",
          };
        }
      }
    );

    // Get contact video
    this.toolExecutor.registerTool(
      defineTool(
        "get_contact_video",
        "Get video info",
        { phoneNumber: { type: "string" } },
        ["phoneNumber"]
      ),
      async (input, context) => {
        const contact = context.contact;

        if (!contact) {
          return { found: false, error: "Contact not found" };
        }

        return {
          found: true,
          hasVideo: !!contact.heygen_video_id,
          videoId: contact.heygen_video_id,
          videoUrl: contact.heygen_video_url,
        };
      }
    );
  }

  /**
   * Determine if handoff to another agent is needed
   */
  protected determineNextAgent(): AgentType | undefined {
    // Video is typically a leaf operation
    return undefined;
  }

  /**
   * Check if this agent can handle a given action
   */
  canHandle(action: string): boolean {
    const supportedActions = [
      "generate_video",
      "check_video_status",
      "get_video",
    ];
    return supportedActions.includes(action);
  }
}

// Register the factory for this agent
registerAgentFactory("video", () => new VideoAgent());

export default VideoAgent;

File: crewai/src/swarm/agents/video_agent.py
"""
Video Agent - HeyGen video generation

Subscribes to: video.requested, video.script_ready
Publishes: video.started, video.completed
"""

import os
import logging
from typing import List

import httpx

from ..agent_runner import AutonomousAgent
from ..events import SwarmEvent, EventType
from ..eventbus import EventBus
from ..blackboard import Blackboard, ContactState

logger = logging.getLogger(__name__)

SYSTEM_OWNER_ID = (
    os.getenv("NEXTHELLO_SYSTEM_OWNER_ID")
    or os.getenv("NEXTHELLO_DEFAULT_OWNER_ID")
    or "askmikeai@gmail.com"
)


class VideoAgent(AutonomousAgent):
    """
    Autonomous agent for video generation using HeyGen.

    Generates personalized AI avatar videos for qualified leads.

    Autonomy logic:
    - Only acts for hot/warm leads
    - Skips if video already exists
    - Requires script (from personalization agent or event payload)
    """

    @property
    def name(self) -> str:
        return "video"

    @property
    def subscribed_events(self) -> List[EventType]:
        return [
            EventType.VIDEO_REQUESTED,
            EventType.VIDEO_SCRIPT_READY,
        ]

    async def should_act(
        self,
        event: SwarmEvent,
        contact: ContactState,
    ) -> bool:
        """
        Decide if we should generate a video.

        Conditions for acting:
        1. Contact doesn't already have a video
        2. Contact is hot or warm tier
        3. Script is available (from payload or contact state)
        """
        # Skip if video already exists
        if contact.heygen_video_url:
            logger.debug(f"[{self.name}] Skipping - video already exists")
            return False

        # Skip if video already in progress
        if contact.heygen_video_id:
            logger.debug(f"[{self.name}] Skipping - video generation in progress")
            return False

        is_welcome_video = bool(event.payload.get("welcome_video"))

        # Only generate for qualified leads unless this is a first-contact welcome video
        tier = contact.qualification_tier
        if not is_welcome_video and tier not in ["hot", "warm"]:
            logger.debug(f"[{self.name}] Skipping - tier {tier} not qualified for video")
            return False

        # Need a script
        script = event.payload.get("script") or contact.video_script
        if not script:
            logger.debug(f"[{self.name}] Skipping - no script available")
            return False

        return True

    async def execute(
        self,
        event: SwarmEvent,
        contact: ContactState,
    ) -> List[SwarmEvent]:
        """
        Execute video generation via HeyGen.
        """
        result_events = []

        # Get script from payload or contact state
        script = event.payload.get("script") or contact.video_script

        if not script:
            logger.error(f"[{self.name}] No script available for {contact.phone_number}")
            return result_events

        cfg = await self.get_owner_config(event.owner_id)

        try:
            # Generate video
            video_id = await self._generate_video(contact.phone_number, script, cfg)

            if video_id:
                # Update contact with video ID (URL will come via webhook)
                await self.blackboard.update_contact(
                    contact.phone_number,
                    owner_id=event.owner_id,
                    heygen_video_id=video_id,
                )

                # Publish video.started
                result_events.append(
                    event.create_response(
                        event_type=EventType.VIDEO_STARTED,
                        payload={
                            "video_id": video_id,
                            "script_length": len(script),
                        },
                        source_agent=self.name,
                    )
                )

                logger.info(
                    f"[{self.name}] Video generation started for {contact.phone_number}: "
                    f"video_id={video_id}"
                )

        except Exception as e:
            logger.error(f"[{self.name}] Video generation failed: {e}")

        return result_events

    async def _generate_video(
        self,
        phone_number: str,
        script: str,
        cfg=None,
    ) -> str | None:
        """
        Generate a HeyGen video.

        Returns video_id or None on failure.
        """
        hg = cfg.heygen if cfg else {}
        beh = cfg.behavior if cfg else {}
        api_key = hg.get("api_key") or os.getenv("HEYGEN_API_KEY")
        if not api_key:
            logger.warning("HEYGEN_API_KEY not configured")
            return None

        avatar_id = hg.get("avatar_id") or os.getenv("HEYGEN_AVATAR_ID")
        voice_id = hg.get("voice_id") or os.getenv("HEYGEN_VOICE_ID")

        if not avatar_id or not voice_id:
            logger.warning("HEYGEN_AVATAR_ID and HEYGEN_VOICE_ID must be set")
            return None

        # Build webhook URL
        webhook_url = None
        base_url = beh.get("webhook_base_url") or os.getenv("WEBHOOK_BASE_URL")
        if base_url:
            webhook_url = f"{base_url}/webhooks/networking-event/heygen"

        request_body = {
            "video_inputs": [
                {
                    "character": {
                        "type": "avatar",
                        "avatar_id": avatar_id,
                        "avatar_style": "normal",
                    },
                    "voice": {
                        "type": "text",
                        "voice_id": voice_id,
                        "input_text": script,
                    },
                }
            ],
            "dimension": {"width": 1280, "height": 720},
            "callback_id": phone_number,
        }

        if webhook_url:
            request_body["callback_url"] = webhook_url

        try:
            async with httpx.AsyncClient(timeout=60.0) as client:
                response = await client.post(
                    "https://api.heygen.com/v2/video/generate",
                    headers={
                        "Content-Type": "application/json",
                        "X-Api-Key": api_key,
                    },
                    json=request_body,
                )

                if response.status_code != 200:
                    logger.error(f"[{self.name}] HeyGen error: {response.status_code}")
                    return None

                data = response.json()

                if data.get("error"):
                    logger.error(f"[{self.name}] HeyGen error: {data['error']}")
                    return None

                return data.get("data", {}).get("video_id")

        except httpx.TimeoutException:
            logger.error(f"[{self.name}] HeyGen request timed out")
            return None
        except Exception as e:
            logger.error(f"[{self.name}] HeyGen error: {e}")
            return None

    async def handle_webhook(
        self,
        video_id: str,
        video_url: str,
        callback_id: str,
        owner_id: str = SYSTEM_OWNER_ID,
    ) -> List[SwarmEvent]:
        """
        Handle HeyGen webhook callback when video is ready.

        This should be called from the webhook endpoint.
        """
        # Update contact with video URL
        await self.blackboard.update_contact(
            callback_id,
            owner_id=owner_id,
            heygen_video_url=video_url,
        )

        # Create and publish video.completed event
        event = SwarmEvent(
            event_type=EventType.VIDEO_COMPLETED,
            contact_id=callback_id,
            owner_id=owner_id,
            payload={
                "video_id": video_id,
                "video_url": video_url,
            },
            source_agent=self.name,
        )

        await self.eventbus.publish(event)

        logger.info(f"[{self.name}] Video completed for {callback_id}: {video_url}")

        return [event]
